src/Language/PureScript/Kinds.hs

```diff
@@ -1,109 +1,164 @@
 -- write it out manually so that we can define a parser which is
 -- backwards-compatible.
-instance A.ToJSON Kind where
+instance A.ToJSON a => A.ToJSON (Kind a) where
   toJSON kind = case kind of
-    KUnknown i ->
-      obj "KUnknown" i
-    Row k ->
-      obj "Row" k
-    FunKind k1 k2 ->
-      obj "FunKind" [k1, k2]
-    NamedKind n ->
-      obj "NamedKind" n
+    KUnknown a i ->
+      obj "KUnknown" a i
+    Row a k ->
+      obj "Row" a k
+    FunKind a k1 k2 ->
+      obj "FunKind" a [k1, k2]
+    NamedKind a n ->
+      obj "NamedKind" a n
     where
-    obj :: A.ToJSON a => Text -> a -> A.Value
-    obj tag contents =
-      A.object [ "tag" .= tag, "contents" .= contents ]
+    obj :: A.ToJSON b => Text -> a -> b -> A.Value
+    obj tag ann contents =
+      A.object [ "tag" .= tag, "annotation" .= ann, "contents" .= contents ]
 
 -- This is equivalent to the derived Aeson FromJSON instance, except that it
 -- also handles JSON generated by compilers up to 0.10.3 and maps them to the
 -- new representations (i.e. NamedKinds which are defined in the Prim module).
-kindFromJSON :: Parse Text Kind
-kindFromJSON = do
+kindFromJSON :: Parse Text a -> Parse Text a -> Parse Text (Kind a)
+kindFromJSON defaultAnn annFromJSON = fix $ \go -> do
   t <- key "tag" asText
+  let annFromJSON' = key "annotation" annFromJSON <|> defaultAnn
   case t of
     "KUnknown" ->
-      KUnknown <$> key "contents" (nth 0 asIntegral)
+      KUnknown <$> annFromJSON' <*> key "contents" (nth 0 asIntegral)
     "Star" ->
-      pure kindType
+      kindType <$> defaultAnn
     "Row" ->
-      Row <$> key "contents" kindFromJSON
+      Row <$> annFromJSON' <*> key "contents" go
     "FunKind" ->
       let
-        kindAt n = key "contents" (nth n kindFromJSON)
+        kindAt n = key "contents" (nth n go)
       in
-        FunKind <$> kindAt 0 <*> kindAt 1
+        FunKind <$> annFromJSON' <*> kindAt 0 <*> kindAt 1
     "Symbol" ->
-      pure kindSymbol
+      kindSymbol <$> defaultAnn
     "NamedKind" ->
-      NamedKind <$> key "contents" fromAesonParser
+      NamedKind <$> annFromJSON' <*> key "contents" fromAesonParser
     other ->
       throwCustomError (T.append "Unrecognised tag: " other)
 
   where
   -- The following are copied from Environment and reimplemented to avoid
   -- circular dependencies.
-  primName :: Text -> Qualified (ProperName a)
+  primName :: Text -> Qualified (ProperName b)
   primName = Qualified (Just $ ModuleName [ProperName C.prim]) . ProperName
 
-  primKind :: Text -> Kind
-  primKind = NamedKind . primName
+  primKind = flip NamedKind . primName
 
   kindType = primKind C.typ
   kindSymbol = primKind C.symbol
 
-instance A.FromJSON Kind where
-  parseJSON = toAesonParser id kindFromJSON
+instance {-# OVERLAPPING #-} A.FromJSON (Kind SourceAnn) where
+  parseJSON = toAesonParser id (kindFromJSON (pure NullSourceAnn) fromAesonParser)
+
+instance {-# OVERLAPPING #-} A.FromJSON a => A.FromJSON (Kind a) where
```

Comment:
Are we using both of these instances?

Reply:
(perhaps a comment to explain what's going on here would help)

Reply:
Ohh - this hadn't occurred to me until now; we'll have to clear out the Pursuit database anyway because the Type codec will be changing, right?

Reply:
Unfortunately, yes. We can perhaps stop deriving Type, if we want to avoid it. The docs are using `()` for the annotation, so it's "unused", but it will still try to parse the field with the derived Type codec.

Reply:
I had intended to come back to Type and put together manual json codecs for it, incidentally, although obviously I haven't managed that yet. Ideally none of our codecs would be derived, I think, at least not for anything appearing in json for docs. See also https://github.com/purescript/purescript/issues/2655

Reply:
How about leaving this for now, and making a new issue to remind us to come back to it before we make another release, to see if we can move away from derived codecs for Type and make this non-breaking from that perspective?